From aed6de59e0ac218b4f95a5944752767774e0ddde Mon Sep 17 00:00:00 2001
From: Hari Chandrakanthan <quic_haric@quicinc.com>
Date: Mon, 20 Nov 2023 16:49:26 +0530
Subject: [PATCH] wifi: ath12k: update peer id in dp monitor path

mac80211 receives a data frame from an associated station with invalid link_id
and link_valid and goes for WARN_ON.

This is because, in monitor path, the link_valid and link_id is not filled with
proper values by ath12k driver in ath12k_dp_mon_rx_deliver_msdu().

The link_valid and link_id is not updated in ath12k_dp_mon_rx_deliver_msdu()
because peer find fails in ath12k_dp_mon_rx_deliver_msdu().

peer find fails in ath12k_dp_mon_rx_deliver_msdu() due to rxcb->peer_id
not being updated with proper value and due to this link_id and link_valid
does not get updated.

Fix this by finding the peer using ath12k_peer_find_by_id() and ppduinfo->peer_id.
ppduinfo->peer_id holds the valid peer id information.

Below is the call trace observed during warn_on
Call trace:
 ieee80211_rx_nss_notify_4addr+0x128/0x32c [mac80211]
 ieee80211_rx_list+0x944/0xd20 [mac80211]
 ieee80211_rx_napi+0x54/0xc8 [mac80211]
 ath12k_pci_exit+0x1280/0x1c0c [ath12k]
 ath12k_dp_mon_rx_parse_mon_status+0x9c/0xa8 [ath12k]
 ath12k_dp_mon_srng_process+0x180/0x528 [ath12k]
 ath12k_dp_mon_process_ring+0x98/0x16d4 [ath12k]
 ath12k_dp_service_srng+0x1f0/0x2b8 [ath12k]
 ath12k_pcic_dummy_irq_handler+0x288/0x474 [ath12k]
 __napi_poll.constprop.0+0x30/0x160
 net_rx_action+0x154/0x264
 _stext+0x1c4/0x228
 ____do_softirq+0xc/0x14
 call_on_irq_stack+0x24/0x34
 do_softirq_own_stack+0x18/0x20
 __irq_exit_rcu+0x94/0xe0
 irq_exit_rcu+0xc/0x14
 el1_interrupt+0x34/0x5c
 el1h_64_irq_handler+0x14/0x1c
 el1h_64_irq+0x68/0x6c
 arch_cpu_idle+0x14/0x20
 default_idle_call+0x28/0x34
 do_idle+0x12c/0x250
 cpu_startup_entry+0x20/0x24
 __cpu_disable+0x0/0xb0
 __secondary_switched+0x64/0x68
Code: b9407006 b0000280 9108fc00 95ba22d5 (d4210000)
---[ end trace 0000000000000000 ]---

Signed-off-by: Hari Chandrakanthan <quic_haric@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_mon.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/net/wireless/ath/ath12k/dp_mon.c b/drivers/net/wireless/ath/ath12k/dp_mon.c
index d2a2dc7..1d33bcc 100644
--- a/drivers/net/wireless/ath/ath12k/dp_mon.c
+++ b/drivers/net/wireless/ath/ath12k/dp_mon.c
@@ -2176,7 +2176,7 @@ static void ath12k_dp_mon_rx_deliver_msdu(struct ath12k *ar, struct napi_struct
 	if (!(status->flag & RX_FLAG_ONLY_MONITOR))
 		decap = ppduinfo->cmn_mpdu_info.mon_mpdu[mpdu_idx].msdu_info[0].decap_format;
 	spin_lock_bh(&ar->ab->base_lock);
-	peer = ath12k_dp_rx_h_find_peer(ar->ab, msdu, rx_info);
+	peer = ath12k_peer_find_by_id(ar->ab, ppduinfo->peer_id);
 	if (peer && peer->sta) {
 		pubsta = peer->sta;
 		if (pubsta->valid_links) {
-- 
2.34.1

